Enable muon physics and add user-defined list of offloaded particles

stognini opened the pull request:
This PR enables muon physics by default and adds the new `SetupOptions::offload_particles` which allows specifying a subset of particles to be offloaded to Celeritas via the Tracking Manager Integration.

This new feature allows for easier individual physics validations in Geant4-Celeritas applications.

stognini committed:
Add deactivated() option to GeantPhysicsOptions

stognini committed:
Allow subset of particles to be configured for offloading in TrackingManagerConstructor; Fix GeantUtils msg typo

stognini committed:
Draft user-defined subset of particles to be offloaded

stognini committed:
Only log warning if not all offload particles are attached; Correctly import extended EM particles/processes

stognini committed:
Fix/improve particle offload in tracking manager constructor

stognini committed:
Merge branch 'develop' into update-offload

stognini committed:
Enable GeantPhysicsOptions::muon by default

stognini committed:
Disable muon msc by default; fix tests

stognini committed:
Cleanup

sethrj commented on src/accel/TrackingManagerIntegration.cc:
```diff
@@ -111,14 +111,10 @@ void verify_tracking_managers(Span<G4PD const* const> expected,
+    CELER_LOG(warning)
+        << "not all particles from TrackingManagerConstructor are active in "
+           "Celeritas: missing "
+        << join(missing.begin(), missing.end(), ", ", printable_pd);
```
Are you sure these changes are correct? Celeritas can't do anything with track types that it doesn't know about. Also the tracking managers should definitely be attached right? I don't understand this change at all.

stognini replied:
Oh you have a point. That was a trailing change from my initial tests before I had the user list. We should only have managers to the offloaded particles, so I have to update the `TrackingManagerIntegration::BeginOfRunAction`:

```cpp
verify_tracking_managers(
                make_span(TrackingManagerConstructor::OffloadParticles()), // <-- Use user-defined list here 
                make_span(singleton.shared_params().OffloadParticles()),
                singleton.shared_params(),
                singleton.local_transporter())
```

sethrj commented on src/accel/TrackingManagerIntegration.cc:
```diff
@@ -51,7 +51,7 @@ void verify_tracking_managers(Span<G4PD const* const> expected,
     auto log_tm_failure = [&all_attached_correctly](G4PD const* p) {
         all_attached_correctly = false;
-        auto msg = CELER_LOG(error);
+        auto msg = CELER_LOG(warning);
```
This also seems wrong; how can you offload a particle to celeritas if the tracking manager isn't attached when using tracking manager integration?

sethrj commented on src/accel/TrackingManagerConstructor.cc:
```diff
@@ -98,14 +162,16 @@ void TrackingManagerConstructor::ConstructProcess()
         // (Note that it is leaked in Geant4 11.0 and 11.1 for MT mode.)
         p->SetTrackingManager(manager ? manager.release() : manager_ptr);
+        CELER_LOG(info) << "Added particle \"" << p->GetParticleName()
+                        << "\" to Tracking Manager ";
```
I think we want one log message, not one per particle:
```c++
}
CELER_LOG(info) << "Built Celeritas tracking managers for "
    << join(offload_particles_.begin(), offload_particles_.end(), ", ",
               [](G4ParticleDefinition const* pd)
               {
                   return p->GetParticleName();
               });
```

sethrj commented on src/accel/TrackingManagerConstructor.cc:
```diff
@@ -70,6 +118,22 @@ TrackingManagerConstructor::TrackingManagerConstructor(
+{
+    auto& singleton = detail::IntegrationSingleton::instance();
+    auto const& user_offload = singleton.setup_options().offload_particles;
+    // Construction of particles happens at offload_particles_ assignment
```
What does this comment mean?

stognini replied:
Since `G4Particle::Definition()` is called when setting up the `offload_particles_` list, a singleton is instantiated for each of those particles in the G4 application.

sethrj replied:
Oh, ok the `G4Particle` particle singletons... there are way too many particles types and instances and such to keep track of 😉

stognini replied:
I'll reword that comment, but thought was useful to have something there, because yeah, it's far from obvious that that call would technically construct the particles (as the `ConstructParticle()` method implies).

sethrj replied:
Totally agree...

sethrj commented on src/accel/TrackingManagerConstructor.cc:
```diff
@@ -70,6 +118,22 @@ TrackingManagerConstructor::TrackingManagerConstructor(
+/*!
+ * Construct particles before \c ::ConstructProcess since any
+ * \c G4ParticleDefinition other than ions and shortlived must be created
+ * during Geant4's \c Pre_Init state.
```
Is this because calling `Blah::Definition` "constructs" it and will error if it's called for the first time during `ConstructProcess`?

stognini replied:
Yep. If we do not initialize them during that call, and do it during `ConstructProcess` we get

```
G4ParticleDefinition (other than ions and shortlived) should be created in Pre_Init state - mu-

-------- WWWW ------- G4Exception-START -------- WWWW -------
*** G4Exception : PART101
      issued by : G4ParticleDefintion::G4ParticleDefinition()
G4ParticleDefinition should be created in PreInit state
*** This is just a warning message. ***
-------- WWWW -------- G4Exception-END --------- WWWW -------
```

sethrj commented on src/accel/SetupOptions.cc:
```diff
@@ -255,8 +255,10 @@ inp::FrameworkInput to_inp(SetupOptions const& so)
+    //! \todo: Expose DataSelection to SetupOptions or default to all
+    // This avoids missing a process/particle import from GeantPhysicsOptions
+    result.geant.data_selection.particles = GeantImportDataSelection::em;
+    result.geant.data_selection.processes = GeantImportDataSelection::em;
```
I think we want to rework the `inp::GeantImport` so that the process and particle lists live there. Or does the construction ordering mean that the tracking manager setup happens before the celeritas setup? 🤪

stognini committed:
Fix tests

stognini committed:
Fix TrackingManagerIntegration:
- Revert changes
- Update verify_tracking_managers to load the correct set of offloaded
  particles

stognini committed:
Use `join` in CELER_LOG msg

sethrj commented on src/accel/TrackingManagerConstructor.cc:
```diff
@@ -25,15 +28,88 @@ namespace celeritas
+ * Get the list of user-defined particles that will be offloaded.
+ *
+ * \warning This assumes that the \c TrackingManagerIntegration::SetOptions has
+ * been called.
```
We should dedicate a section of the offloading part of the user manual to discuss the order of operations... maybe a little flow chart for the dependencies of "A must be called before B"? I think it should be set at this point.

stognini committed:
Improve readability; deactivate muon physics by default;

sethrj commented on src/celeritas/ext/GeantPhysicsOptions.hh:
```diff
@@ -200,6 +212,30 @@ struct GeantPhysicsOptions
     GeantOpticalPhysicsOptions optical{
         GeantOpticalPhysicsOptions::deactivated()};
+
+    //! Initialize with no physics
```
I think we've been meaning to move most of that list into a separate "EM physics" struct... wonder if now is the right time to do it?

sethrj replied:
Let's defer.

sethrj commented on src/accel/TrackingManagerConstructor.cc:
```diff
@@ -25,15 +28,95 @@ namespace celeritas
+
+    auto const supported
+        = TrackingManagerConstructor::SupportedOffloadParticles();
+    CELER_VALIDATE(user.size() <= supported.size(),
+                   << "List of particles defined in "
+                      "SetupOptions::offload_particles is larger than the "
+                      "list of supported particles in Celeritas, which is: "
+                   << join(supported.begin(),
+                           supported.end(),
+                           ", ",
+                           [](G4ParticleDefinition const* pd) {
+                               return pd->GetParticleName();
+                           }););
+
```
I think this check is redundant.

sethrj commented on src/accel/TrackingManagerConstructor.hh:
```diff
@@ -50,11 +50,18 @@ class TrackingManagerConstructor final : public G4VPhysicsConstructor
     //!@{
     //! \name Type aliases
     using LocalTransporterFromThread = std::function<LocalTransporter*(int)>;
+    using VecG4PD = std::vector<G4ParticleDefinition*>;
```
Since the ordering for these doesn't matter (I just used a `Span` originally because the data was easily stored as a static array) let's use
```suggestion
    using SetG4PD = std::unordered_set<G4ParticleDefinition*>;
```
and return const references to static SetG4PD. That should also simplify your testing for user-requested particles.

sethrj commented on src/accel/TrackingManagerIntegration.cc:
```diff
@@ -156,9 +158,16 @@ void TrackingManagerIntegration::BeginOfRunAction(G4Run const*)
+        auto const offload_particles
+            = opts.offload_particles.empty()
+                  ? TrackingManagerConstructor::DefaultOffloadParticles()
+                  : TrackingManagerConstructor::UserOffloadParticles();
```
This is pretty awkward since it reconstructs the offload particles again. How about having `TrackingManagerConstructor`'s `ConstructParticle` method setting some static data in `TrackingManagerIntegration` ? It might be better to have a conversation about this.

stognini committed:
Reorganize offload logic:

- SetupOptions takes a vector of G4ParticleDefinition* to be offloaded
- Move supported, default, and loaded offloaded particles to
  IntegrationSingleton
- IntegrationSingleton is responsible for verifying correctness of SetupOptions
  list of offloaded particles when SetupOptions are loaded
- Both TrackingManagerConstructor and SharedParams now assign a correct list of
  offloaded particles
- Since the IntegrationSingleton has the user list we can also assign
  DataSelection flags in inp::FrameworkInput::to_inp(SetupOptions const&),
  which then imports the data correctly in GeantImporter

stognini committed:
Fix typo; minor improvements

stognini committed:
Squash bugs

stognini committed:
Update docs

sethrj commented on src/accel/SharedParams.cc:
```diff
@@ -258,6 +250,20 @@ SharedParams::SharedParams(SetupOptions const& options)
+        // Set up offloaded particles based on user input
+        auto const& user_offload = options.offload_particles;
+        particles_ = user_offload.empty() ? IS::default_offload_particles()
+                                          : user_offload;
```
```suggestion
    if (mode_ == Mode::enabled || mode_ == Mode::kill_offload)
    {
        // Set up offloaded/killed particles based on user input
        auto const& user_offload = options.offload_particles;
        particles_ = user_offload.empty() ? IS::default_offload_particles()
                                          : user_offload;
```

Also maybe SharedParams is a better place to store the hardcoded "supported particles" vetors? We were originally doing that at one point, and it seems like a more accessible+appropriate place to get the list of default+supported.

sethrj commented on src/accel/TrackingManagerConstructor.cc:
```diff
@@ -70,6 +53,20 @@ TrackingManagerConstructor::TrackingManagerConstructor(
+    // Construction of particles happens at offload_particles_ assignment,
+    // since it will instantiate the G4Particle::Definition() singletons
+    offload_particles_
+        = detail::IntegrationSingleton::instance().offloaded_particles();
```
Wait, this should be assigned only in the delegating `TrackingManagerConstructor(TrackingManagerIntegration` constructor, right? Otherwise this class can't be used without the singleton.

sethrj commented on src/accel/detail/IntegrationSingleton.cc:
```diff
@@ -23,6 +29,46 @@ namespace celeritas
+        CELER_VALIDATE(find(pd),
+                       << "Particle \"" << pd->GetParticleName()
+                       << "\" (PDG = " << pd->GetPDGEncoding()
+                       << ") is not available in Celeritas");
```
I forgot about PrintablePD in `geocel/GeantUtils.hh`: use that here.

sethrj reviewed (commented):
One more set of small changes I think? Remember we need to support the case *without* the singleton.

stognini committed:
Address comments

stognini committed:
Merge branch 'develop' into update-offload

sethrj reviewed (approved):
Nice! Just a typo in the doxygen causing failures.

amandalund commented on src/celeritas/ext/GeantPhysicsOptions.hh:
```diff
@@ -53,11 +53,11 @@ enum class RelaxationSelection
+    bool ionization{true};
     //! Enable muon bremsstrahlung
-    bool bremsstrahlung{false};
+    bool bremsstrahlung{true};
```
Can we also enable MSC with the Urban model?

stognini committed:
Enable muon msc urban model by default; fix doc

sethrj commented:
@stognini The final failure is in a diff used to show how to add celeritas in the documentation (it's now wrong!) Could we relax the ordering of `tmi.SetOptions(MakeOptions())`: check the user input particles, and if they're the default ones (or empty) then allow it as usual; if not, then fail and tell the user to move before the tracking constructor?

stognini committed:
Address comments:
- Relax TMI::SetOptions ordering
- Revert trackingmanager-offload example app
- Update interface doc

stognini committed:
Fixup

sethrj committed:
Fix kill-offload particle setting and add validation
